winch/codegen/src/masm.rs

```diff
@@ -199,81 +199,76 @@ pub(crate) enum ShiftKind {
 /// lowering to machine code.
 #[derive(Copy, Clone)]
 pub(crate) enum ExtendKind {
-    Signed(SignedExtend),
-    Unsigned(UnsignedExtend),
+    Signed(Extend<Signed>),
+    Unsigned(Extend<Zero>),
 }
 
 #[derive(Copy, Clone)]
-pub(crate) enum UnsignedExtend {
-    /// 8 to 32 bit unsigned extend.
-    I32Extend8U,
-    /// 16 to 32 bit unsigned extend.
-    I32Extend16U,
-    /// 8 to 64 bit unsigned extend.
-    I64Extend8U,
-    /// 16 to 64 bit unsigned extend.
-    I64Extend16U,
-    /// 32 to 64 bit unsigned extend.
-    I64Extend32U,
+pub(crate) enum Signed {}
+#[derive(Copy, Clone)]
+pub(crate) enum Zero {}
+
+pub(crate) trait ExtendType {}
+
+impl ExtendType for Signed {}
+impl ExtendType for Zero {}
+
+#[derive(Copy, Clone)]
+pub(crate) enum Extend<T: ExtendType> {
+    /// 8 to 32 bit extend.
+    I32Extend8,
+    /// 16 to 32 bit extend.
+    I32Extend16,
+    /// 8 to 64 bit extend.
+    I64Extend8,
+    /// 16 to 64 bit extend.
+    I64Extend16,
+    /// 32 to 64 bit extend.
+    I64Extend32,
+
+    /// Variant to hold the kind of extend marker.
+    ///
+    /// This is `Signed` or `Zero`, that are empty enums, which means that this variant cannot be
+    /// constructed.
+    __Kind(T),
 }
 
-impl From<UnsignedExtend> for ExtendKind {
-    fn from(value: UnsignedExtend) -> Self {
+impl From<Extend<Zero>> for ExtendKind {
+    fn from(value: Extend<Zero>) -> Self {
         ExtendKind::Unsigned(value)
     }
 }
 
-impl UnsignedExtend {
-    pub fn from_bits(&self) -> u8 {
+impl<T: ExtendType> Extend<T> {
+    pub fn src_size(&self) -> OperandSize {
```

Comment:
Small naming convention/suggestion: we usually use `src`, `dst` for explicit moves, could we keep `from_bits` and `to_bits`? Although given that this now returns `OperandSize` it could maybe be `from_size` and `to_size`?

Reply:
No problem. It's just that it kinda overlaps with usual rust naming convention for `from_` and `to_`, which I found a bit confusing. I'll make the change now.